core/src/test/java/com/github/mizool/core/configuration/TestConfig.java

```diff
@@ -0,0 +1,336 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.mizool.core.configuration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+
+import java.math.BigDecimal;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.Period;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import com.github.mizool.core.exception.ConfigurationException;
+import com.google.common.collect.ImmutableList;
+
+public class TestConfig
+{
+    private static final String VALUE = "foo";
+
+    private Properties testProperties;
+
+    @BeforeMethod
+    void setUp()
+    {
+        testProperties = new Properties();
+        testProperties.setProperty("answer", "forty-two");
+        testProperties.setProperty("something", "interesting");
+
+        testProperties.setProperty("a", VALUE);
+        testProperties.setProperty("a.b", VALUE);
+        testProperties.setProperty("a.b.c", VALUE);
+    }
+
+    @Test
+    public void testProperties()
+    {
+        RootNode root = Config.from(testProperties);
+
+        assertThat(root.child("answer").stringValue().obtain()).isEqualTo("forty-two");
+        assertThat(root.child("something").stringValue().obtain()).isEqualTo("interesting");
+    }
+
+    @Test
+    public void testPropertiesWithDefaults()
+    {
+        Properties propertiesWithDefaults = new Properties(testProperties);
+        propertiesWithDefaults.setProperty("greeting", "Hello, World!");
+
+        RootNode root = Config.from(propertiesWithDefaults);
+        assertThat(root.child("answer").stringValue().obtain()).isEqualTo("forty-two");
+        assertThat(root.child("something").stringValue().obtain()).isEqualTo("interesting");
+        assertThat(root.child("greeting").stringValue().obtain()).isEqualTo("Hello, World!");
+    }
+
+    @Test
+    public void testSystemProperties()
+    {
+        RootNode root = Config.systemProperties();
+        String lineSeparator = root.child("line.separator").stringValue().obtain();
+        assertThat(lineSeparator).isNotEmpty();
+    }
+
+    @Test
+    public void testConfigPicksUpPropertiesChangedLater()
+    {
+        Properties properties = new Properties();
+        PropertyNode node = Config.from(properties).child("breaking");
+
+        properties.setProperty("breaking", "news");
+
+        assertThat(node.stringValue().obtain()).isEqualTo("news");
```

Comment:
I'd like to change this to make it even more ballsy: extract a variable for `node.stringValue()` before setting the property, to ensure that even the `Value` class is lazy.

Also you could extend the test to first test that `value.read()` returns an `Optional.absent()`, then setting the property and asserting the `value.obtain()` again.

Reply:
Since there are already tests that ensure the function of `obtain`, I'd recommend going with `read` for the other tests to avoid being distracted by the exception thrown by `obtain`.